File: app/src/main/java/cn/winxo/toolbox/data/dao/AppDatabase.java
package cn.winxo.toolbox.data.dao;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;
import cn.winxo.toolbox.data.entity.local.Task;
import cn.winxo.toolbox.util.Converters;

@Database(entities = { Task.class }, version = 1) @TypeConverters(Converters.class) public abstract class AppDatabase extends RoomDatabase {

  private static volatile AppDatabase instance;
  private static final String name = "winxo.db";

  public static AppDatabase getInstance(Context context) {
    if (instance == null) {
      synchronized (AppDatabase.class) {
        if (instance == null) {
          instance = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, name).build();
        }
      }
    }
    return instance;
  }

  public abstract TaskDao taskDao();
}

File: app/src/main/java/cn/winxo/toolbox/data/dao/AppDatabase.kt
package cn.winxo.toolbox.data.dao

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import cn.winxo.toolbox.data.entity.local.Task
import cn.winxo.toolbox.data.entity.local.Type
import cn.winxo.toolbox.util.Converters

@Database(entities = [Task::class, Type::class], version = 1)
@TypeConverters(Converters::class)
abstract class AppDatabase : RoomDatabase() {

    abstract fun taskDao(): TaskDao

    abstract fun typeDao(): TypeDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null
        private const val name = "winxo.db"

        fun getInstance(context: Context): AppDatabase = INSTANCE ?: synchronized(this) {
            INSTANCE ?: Room.databaseBuilder(context.applicationContext,
                    AppDatabase::class.java, name)
                    .build().also { INSTANCE = it }
        }
    }
}
